elixir/apps/domain/lib/domain/events/decoder.ex

```diff
@@ -0,0 +1,370 @@
+      * `tuple_data` - The data of the inserted tuple.
+      """
+      defstruct [:relation_id, :tuple_data]
+    end
+
+    defmodule Update do
+      @moduledoc """
+      Struct representing the UPDATE message in PostgreSQL's logical decoding output.
+
+      * `relation_id` - The OID of the relation that the tuple was updated in.
+      * `changed_key_tuple_data` - The data of the tuple with the old key values.
+      * `old_tuple_data` - The data of the tuple before the update.
+      * `tuple_data` - The data of the tuple after the update.
+      """
+      defstruct [:relation_id, :changed_key_tuple_data, :old_tuple_data, :tuple_data]
+    end
+
+    defmodule Delete do
+      @moduledoc """
+      Struct representing the DELETE message in PostgreSQL's logical decoding output.
+
+      * `relation_id` - The OID of the relation that the tuple was deleted from.
+      * `changed_key_tuple_data` - The data of the tuple with the old key values.
+      * `old_tuple_data` - The data of the tuple before the delete.
+      """
+      defstruct [:relation_id, :changed_key_tuple_data, :old_tuple_data]
+    end
+
+    defmodule Truncate do
+      @moduledoc """
+      Struct representing the TRUNCATE message in PostgreSQL's logical decoding output.
+
+      * `number_of_relations` - The number of truncated relations.
+      * `options` - Additional options provided when truncating the relations.
+      * `truncated_relations` - List of relations that have been truncated.
+      """
+      defstruct [:number_of_relations, :options, :truncated_relations]
+    end
+
+    defmodule Type do
+      @moduledoc """
+      Struct representing the TYPE message in PostgreSQL's logical decoding output.
+
+      * `id` - The OID of the type.
+      * `namespace` - The namespace of the type.
+      * `name` - The name of the type.
+      """
+      defstruct [:id, :namespace, :name]
+    end
+
+    defmodule Unsupported do
+      @moduledoc """
+      Struct representing an unsupported message in PostgreSQL's logical decoding output.
+
+      * `data` - The raw data of the unsupported message.
+      """
+      defstruct [:data]
+    end
+  end
+
+  require Logger
+
+  @pg_epoch DateTime.from_iso8601("2000-01-01T00:00:00Z")
+
+  alias Messages.{
+    Begin,
+    Commit,
+    Origin,
+    Relation,
+    Relation.Column,
+    Insert,
+    Update,
+    Delete,
+    Truncate,
+    Type,
+    Unsupported
+  }
+
+  alias Domain.Events.OidDatabase
+
+  @doc """
+  Parses logical replication messages from Postgres
+
+  ## Examples
+
+      iex> decode_message(<<73, 0, 0, 96, 0, 78, 0, 2, 116, 0, 0, 0, 3, 98, 97, 122, 116, 0, 0, 0, 3, 53, 54, 48>>)
+      %Domain.Events.Decoder.Messages.Insert{relation_id: 24576, tuple_data: {"baz", "560"}}
+
+  """
+  def decode_message(message) when is_binary(message) do
+    decode_message_impl(message)
+  end
+
+  defp decode_message_impl(<<"B", lsn::binary-8, timestamp::integer-64, xid::integer-32>>) do
```

Comment:
```suggestion
  defp decode_binary_message(<<"B", lsn::binary-8, timestamp::integer-64, xid::integer-32>>) do
```

Perhaps a bit more descriptive?

Reply:
Should there be a codepath where it is not binary?

Reply:
All messages from our slot will be binary.

Reply:
In Elixir, binary and strings are the same thing - this will crash-on-purpose if we receive something weird like a number or something.